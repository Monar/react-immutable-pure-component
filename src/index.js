/*
  Copyright (C) 2017 Piotr Tomasz Monarski.
  Licensed under the MIT License (MIT), see
  https://github.com/Monar/react-immutable-pure-component
*/

import React from 'react';
import { is } from 'immutable';


export class ImmutablePureComponent extends React.Component {

  shouldComponentUpdate(nextProps, nextState = {}) {
    const state = this.state || {};

    return !(this.updateOnProps || Object.keys({...nextProps, ...this.props})).every((p) => is(nextProps[p], this.props[p]))
      || !(this.updateOnStates || Object.keys({ ...nextState, ...state})).every((s) => is(nextState[s], state[s]));
  }
}

export default ImmutablePureComponent;
