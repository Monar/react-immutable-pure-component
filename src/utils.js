import { is, isImmutable, isOrdered } from 'immutable';

const NOT_SET = {};

export function check(config, prev = {}, next = {}, label) {
  let checkItem = createChecker(prev, next, label);
  let checklist = config || Object.keys({ ...next, ...prev });

  return checklist.every(checkItem);
}

function createChecker(prev, next, checkName) {
  return function(name) {
    if (typeof name === 'string') {
      return is(next[name], prev[name]);
    }

    if (!getIn) {
      let value = JSON.stringify(name);
      throw new TypeError(
        `Not supported value "${value}" provided to ${checkName}, try updating immutable to v4`,
      );
    }

    return is(getIn(next, name), getIn(prev, name));
  };
}

function get(collection, key, notSetValue) {
  return isImmutable(collection)
    ? collection.get(key, notSetValue)
    : typeof collection.get === 'function'
    ? collection.get(key, notSetValue)
    : hasOwnProperty.call(collection, key)
    ? collection[key]
    : notSetValue;
}

function coerceKeyPath(keyPath) {
  if (Array.isArray(keyPath) && typeof keyPath !== 'string') {
    return keyPath;
  }
  if (isOrdered(keyPath)) {
    return keyPath.toArray();
  }
  throw new TypeError(
    'Invalid keyPath: expected Ordered Collection or Array: ' + keyPath,
  );
}

function getIn(collection, searchKeyPath, notSetValue) {
  const keyPath = coerceKeyPath(searchKeyPath);
  let i = 0;
  while (i !== keyPath.length) {
    collection = get(collection, keyPath[i++], NOT_SET);
    if (collection === NOT_SET) {
      return notSetValue;
    }
  }
  return collection;
}
