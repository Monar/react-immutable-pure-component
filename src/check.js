import { is } from 'immutable';
import { getIn } from './utils';

export function check(config, prev = {}, next = {}, label) {
  let checkItem = createChecker(prev, next, label);
  let checklist = config || Object.keys({ ...next, ...prev });

  return checklist.every(checkItem);
}

function createChecker(prev, next) {
  return function(name) {
    if (typeof name === 'string') {
      return is(next[name], prev[name]);
    } else if (Array.isArray(name)) {
      return is(getIn(next, name), getIn(prev, name));
    }

    throw new TypeError('Invalid key: expected Array or string: ' + name);
  };
}
