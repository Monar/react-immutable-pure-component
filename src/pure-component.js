import React from 'react';
import { check } from './utils';

export class ImmutablePureComponent extends React.Component {
  shouldComponentUpdate(nextProps, nextState = {}) {
    return (
      !check(this.updateOnProps, this.props, nextProps, 'updateOnProps') ||
      !check(this.updateOnStates, this.state, nextState, 'updateOnStates')
    );
  }
}
