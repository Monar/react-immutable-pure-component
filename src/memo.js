import React from 'react';
import { check } from './check';

export function immutableMemo(Component, updateOnProps) {
  return React.memo(Component, (prev, next) =>
    check(updateOnProps, prev, next, 'immutableMemo'),
  );
}
