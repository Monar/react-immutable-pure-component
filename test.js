/* eslint-env node, jest */
import { Map, List, fromJS } from 'immutable';
import { ImmutablePureComponent } from './src';

const testProps = {
  someMap: Map({ key: 'value', deep: Map({ fuz: 'buzz' }) }),
  someList: List([1, 'elo', 2]),
  mixed: [{ test: true }, Map({ key: List(['1', 1, { y: 'z' }]) })],
  name: 'superComponent',
  other: 123,
  isOk: true,
};

const testState = {
  someMap: Map({ key: 'value', deep: Map({ fuz: 'buzz' }) }),
  someList: List([1, 'elo', 2]),
  mixed: [{ test: true }, Map({ key: List(['1', 1, { y: 'z' }]) })],
  name: 'superComponent',
  other: 123,
  isOk: true,
};

const getTestComponent = (props = {}, state) => {
  const Component = new ImmutablePureComponent();
  Component.props = props;
  Component.state = state;
  return Component;
};

describe('should not update', () => {
  it('empty', () => {
    const Component = getTestComponent({}, undefined);

    const result = Component.shouldComponentUpdate({}, undefined);
    expect(result).toBe(false);
  });

  it('same instance of props, no state', () => {
    const Component = getTestComponent(testProps, undefined);
    const result = Component.shouldComponentUpdate(testProps, undefined);
    expect(result).toBe(false);
  });

  it('same instance of state no props', () => {
    const Component = getTestComponent({}, testState);
    const result = Component.shouldComponentUpdate({}, testState);
    expect(result).toBe(false);
  });

  it('same instance of state and props', () => {
    const Component = getTestComponent(testProps, testState);
    const result = Component.shouldComponentUpdate(testProps, testState);
    expect(result).toBe(false);
  });

  it('same props and state but new instances', () => {
    const props = {
      someMap: Map({ key: 'value' }),
      someList: List([1, 'elo', 2]),
      name: 'superComponent',
      other: 123,
      isOk: true,
    };
    const state = {
      someMap: Map({ key: 'value' }),
      someList: List([1, 'elo', 2]),
      name: 'superComponent',
      other: 123,
      isOk: true,
    };
    const newProps = {
      someMap: Map({ key: 'value' }),
      someList: List([1, 'elo', 2]),
      name: 'superComponent',
      other: 123,
      isOk: true,
    };
    const newState = {
      someMap: Map({ key: 'value' }),
      someList: List([1, 'elo', 2]),
      name: 'superComponent',
      other: 123,
      isOk: true,
    };
    const Component = getTestComponent(props, state);
    const result = Component.shouldComponentUpdate(newProps, newState);
    expect(result).toBe(false);
  });

  it('same instance of object in props and state', () => {
    const test = { fis: 'bass' };
    const Component = getTestComponent({ test }, { test });
    const result = Component.shouldComponentUpdate({ test }, { test });
    expect(result).toBe(false);
  });

  it('same instance of object in props and state, but modified', () => {
    const test = { fis: 'bass' };
    const Component = getTestComponent({ test }, { test });

    test['new'] = 'new';

    const result = Component.shouldComponentUpdate({ test }, { test });
    expect(result).toBe(false);
  });

  it('same instance of function in props and state', () => {
    const test = { fis: () => {} };
    const Component = getTestComponent({ test }, { test });
    const result = Component.shouldComponentUpdate({ test }, { test });
    expect(result).toBe(false);
  });
});

describe('should update', () => {
  it('when add new property', () => {
    const Component = getTestComponent({}, undefined);

    const result = Component.shouldComponentUpdate({ elo: 'elo' }, undefined);
    expect(result).toBe(true);
  });

  it('when add new state', () => {
    const Component = getTestComponent({}, undefined);

    const result = Component.shouldComponentUpdate({}, { elo: 'elo' });
    expect(result).toBe(true);
  });

  it('when remove property', () => {
    const Component = getTestComponent({ elo: 'elo' }, undefined);

    const result = Component.shouldComponentUpdate({}, undefined);
    expect(result).toBe(true);
  });

  it('when remove state', () => {
    const Component = getTestComponent({}, { elo: 'elo' });

    const result = Component.shouldComponentUpdate({}, {});
    expect(result).toBe(true);
  });

  it('when modify any of the properties', () => {
    const Component = getTestComponent(testProps);

    expect(
      Component.shouldComponentUpdate({
        ...testProps,
        someMap: Map({ key: 'key' }),
      }),
    ).toBe(true, 'Map');
    expect(
      Component.shouldComponentUpdate({ ...testProps, someList: List([1, 2]) }),
    ).toBe(true, 'List');
    expect(
      Component.shouldComponentUpdate({ ...testProps, name: 'otherName' }),
    ).toBe(true, 'string');
    expect(Component.shouldComponentUpdate({ ...testProps, other: 321 })).toBe(
      true,
      'number',
    );
    expect(Component.shouldComponentUpdate({ ...testProps, isOk: false })).toBe(
      true,
      'boolean',
    );
  });

  it('new instance of object in props', () => {
    const test = { fis: 'bass' };
    const Component = getTestComponent({ test });

    const newTest = { fis: 'bass' };
    const result = Component.shouldComponentUpdate({ test: newTest });
    expect(result).toBe(true);
  });

  it('new instance of object in state', () => {
    const test = { fis: 'bass' };
    const Component = getTestComponent({}, { test });

    const newTest = { fis: 'bass' };
    const result = Component.shouldComponentUpdate({}, { test: newTest });
    expect(result).toBe(true);
  });

  it('new instance of function in props and state', () => {
    const test = { fis: () => {} };
    const Component = getTestComponent({ test }, { test });
    const result = Component.shouldComponentUpdate(
      { test: () => {} },
      { test: () => {} },
    );
    expect(result).toBe(true);
  });
});

describe('updateOnProps', () => {
  it('when empty list, and props change', () => {
    const Component = getTestComponent({});
    Component.updateOnProps = [];

    const result = Component.shouldComponentUpdate({ elo: 'elo' });
    expect(result).toBe(false);
  });

  it('when empty list, and state change', () => {
    const Component = getTestComponent({});
    Component.updateOnProps = [];

    const result = Component.shouldComponentUpdate({}, { elo: 'elo' });
    expect(result).toBe(true);
  });

  it('when defined and other prop changes', () => {
    const Component = getTestComponent({ elo: 'elo', isOk: false });
    Component.updateOnProps = ['elo', 'fis', 'buss'];

    const result = Component.shouldComponentUpdate({ elo: 'elo', isOk: true });
    expect(result).toBe(false);
  });

  it('when defined and prop changes', () => {
    const Component = getTestComponent({ elo: 'elo', isOk: false });
    Component.updateOnProps = ['elo', 'fis', 'buss'];

    const result = Component.shouldComponentUpdate({
      elo: 'elo2',
      isOk: false,
    });
    expect(result).toBe(true);
  });

  it('when given non existing prop should not update', () => {
    const Component = getTestComponent({});
    Component.updateOnProps = ['ghost'];

    const result = Component.shouldComponentUpdate({});
    expect(result).toBe(false);
  });
});

describe('updateOnProps deepCheck', () => {
  it('should throw when prop name is not a string and there is not getIn', () => {
    const Component = getTestComponent({ a: 1 });
    Component.updateOnProps = [[]];

    function getShouldComponentUpdateInNewScope() {
      // eslint-disable-next-line
      let x = Component.shouldComponentUpdate;
      return function(...args) {
        x.getIn = undefined;
        return x(...args);
      };
    }

    let w = getShouldComponentUpdateInNewScope();

    expect(() => w.call(Component, { a: 1 })).toThrowErrorMatchingSnapshot();
  });

  it('only empty array "get props" with same in to instances', () => {
    const Component = getTestComponent({ a: 1 });
    Component.updateOnProps = [[]];

    const result = Component.shouldComponentUpdate({ a: 1 });
    expect(result).toBe(true);
  });

  it('only empty array "get props" with same instances', () => {
    let props = { a: 1 };
    const Component = getTestComponent(props);
    Component.updateOnProps = [[]];

    const result = Component.shouldComponentUpdate(props);
    expect(result).toBe(false);
  });

  it('basic test with mutable data', () => {
    let props = { a: { b: ['0'] } };
    let testProps = { a: { b: ['2'] } };
    const Component = getTestComponent(props);
    Component.updateOnProps = [['a', 'b', 0]];

    const result = Component.shouldComponentUpdate(testProps);
    expect(result).toBe(true);
  });

  it('basic test with immutable data', () => {
    let props = { a: fromJS({ b: ['0'] }) };
    let testProps = { a: fromJS({ b: ['2'] }) };
    const Component = getTestComponent(props);
    Component.updateOnProps = [['a', 'b', 0]];

    const result = Component.shouldComponentUpdate(testProps);
    expect(result).toBe(true);
  });

  it('not existing path', () => {
    let props = { a: { b: ['0'] } };
    let testProps = { a: { b: ['2'] } };
    const Component = getTestComponent(props);
    Component.updateOnProps = [['a', 'c']];

    const result = Component.shouldComponentUpdate(testProps);
    expect(result).toBe(false);
  });
});

describe('updateOnStates', () => {
  it('when empty list, and state change', () => {
    const Component = getTestComponent({});
    Component.updateOnStates = [];

    const result = Component.shouldComponentUpdate({}, { elo: 'elo' });
    expect(result).toBe(false);
  });

  it('when empty list, and prop change', () => {
    const Component = getTestComponent({});
    Component.updateOnStates = [];

    const result = Component.shouldComponentUpdate({ elo: 'elo' }, {});
    expect(result).toBe(true);
  });

  it('when defined and other state changes', () => {
    const Component = getTestComponent({}, { elo: 'elo', isOk: false });
    Component.updateOnStates = ['elo', 'fis', 'buss'];

    const result = Component.shouldComponentUpdate(
      {},
      { elo: 'elo', isOk: true },
    );
    expect(result).toBe(false);
  });

  it('when defined and state changes', () => {
    const Component = getTestComponent({}, { elo: 'elo', isOk: false });
    Component.updateOnProps = ['elo', 'fis', 'buss'];

    const result = Component.shouldComponentUpdate(
      {},
      { elo: 'elo2', isOk: false },
    );
    expect(result).toBe(true);
  });
});

describe('updateOnStates deepCheck', () => {
  it('only empty array "get states" with same in to instances', () => {
    const Component = getTestComponent({}, { a: 1 });
    Component.updateOnStates = [[]];

    const result = Component.shouldComponentUpdate({}, { a: 1 });
    expect(result).toBe(true);
  });

  it('only empty array "get state" with same instances', () => {
    let state = { a: 1 };
    const Component = getTestComponent({}, state);
    Component.updateOnStates = [[]];

    const result = Component.shouldComponentUpdate({}, state);
    expect(result).toBe(false);
  });

  it('basic test with mutable data', () => {
    let state = { a: { b: ['0'] } };
    let testState = { a: { b: ['2'] } };
    const Component = getTestComponent({}, state);
    Component.updateOnStates = [['a', 'b', 0]];

    const result = Component.shouldComponentUpdate({}, testState);
    expect(result).toBe(true);
  });

  it('basic test with immutable data', () => {
    let state = { a: fromJS({ b: ['0'] }) };
    let testState = { a: fromJS({ b: ['2'] }) };
    const Component = getTestComponent({}, state);
    Component.updateOnStates = [['a', 'b', 0]];

    const result = Component.shouldComponentUpdate({}, testState);
    expect(result).toBe(true);
  });

  it('not existing path', () => {
    let state = { a: { b: ['0'] } };
    let testState = { a: { b: ['2'] } };
    const Component = getTestComponent({}, state);
    Component.updateOnStates = [['a', 'c']];

    const result = Component.shouldComponentUpdate({}, testState);
    expect(result).toBe(false);
  });
});

describe('updateOnStates and updateOnProps', () => {
  it('have some values', () => {
    const Component = getTestComponent({ p: 1, x: 1 }, { s: 1, x: 1 });
    Component.updateOnProps = ['p'];
    Component.updateOnStates = ['s'];

    expect(Component.shouldComponentUpdate({ p: 1 }, { s: 1 })).toBe(false);
    expect(Component.shouldComponentUpdate({ p: 2 }, { s: 1 })).toBe(true);
    expect(Component.shouldComponentUpdate({ p: 1 }, { s: 2 })).toBe(true);
    expect(Component.shouldComponentUpdate({ p: 2 }, { s: 2 })).toBe(true);
  });

  it('are both default', () => {
    const Component = getTestComponent({ p: 1, x: 1 }, { s: 1, x: 2 });

    expect(
      Component.shouldComponentUpdate({ p: 1, x: 1 }, { s: 1, x: 2 }),
    ).toBe(false);
    expect(Component.shouldComponentUpdate({ p: 1 }, { s: 1 })).toBe(true);
    expect(Component.shouldComponentUpdate({ p: 1, x: 1 }, { s: 2 })).toBe(
      true,
    );
    expect(Component.shouldComponentUpdate({ p: 2 }, { s: 1, x: 2 })).toBe(
      true,
    );
  });

  it('are both empty', () => {
    const Component = getTestComponent({ p: 1, x: 1 }, { s: 1, x: 2 });
    Component.updateOnProps = [];
    Component.updateOnStates = [];

    expect(
      Component.shouldComponentUpdate({ p: 1, x: 1 }, { s: 1, x: 2 }),
    ).toBe(false);
    expect(Component.shouldComponentUpdate({ p: 1 }, { s: 1 })).toBe(false);
    expect(Component.shouldComponentUpdate({ p: 1, x: 1 }, { s: 2 })).toBe(
      false,
    );
    expect(Component.shouldComponentUpdate({ p: 2 }, { s: 1, x: 2 })).toBe(
      false,
    );
  });
});
