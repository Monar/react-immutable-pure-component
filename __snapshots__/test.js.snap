// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateOnProps deepCheck should throw when prop name is not a string and there is not getIn 1`] = `"Cannot read property 'state' of undefined"`;
